prepreduce/prepreduce.py

```diff
@@ -0,0 +1,395 @@
+    ic(local_path_for_dot_i_file)
+
+    if (loc := local_path_for_dot_i_file).exists():
+        new_name = loc.with_name(".old." + loc.stem + loc.suffix)
+        warn(f"{loc} already exists. Renaming it to {new_name}")
+        os.rename(local_path_for_dot_i_file, output_dir / new_name)
+    info(
+        "Running cc invocation with -E to generate preprocessed .i file... "
+        f"Will be saved to {local_path_for_dot_i_file}"
+    )
+
+    cc[output_target_idx] = str(local_path_for_dot_i_file)
+
+    make_result = subprocess.run(
+        cc, text=True, cwd=path_to_linux, stderr=subprocess.STDOUT
+    )
+    ic(make_result)
+
+    if make_result.returncode != 0:
+        error(
+            f"""
+              Making the .i target file failed. Here's some diagnostics:
+              {make_result.stdout=}
+              {cc=}
+              {local_path_for_dot_i_file=}
+              {make_result.returncode}
+              """
+        )
+
+    success(
+        f"Made the .i target file successfully. It's located here: {local_path_for_dot_i_file}"
+    )
+
+    return local_path_for_dot_i_file
+
+
+def get_compiler_invocation(make_stdout: str) -> list[str]:
+    """
+    Parse out the CC invocation from make's raw output with V=1 on
+    """
+    pattern = r".*-c -o.*"  # HACK: how robust is this pattern?
+    matches = re.findall(pattern, make_stdout)
+    ic(matches)
+    assert (
+        len(matches) == 1
+    ), "Too many (or too few) compiler invocation matches! (not your fault, it's prepreduce's fault)"
+
+    main_invocation = matches[0].split()
+    return main_invocation
+
+
+def get_clean_compiler_invocation(cc_invocation: list[str]) -> list[str]:
+    """
+    Remove all preprocessor flags as well as any relative pathing in invocation
+    """
+    rules = [
+        lambda flag: flag[:2] != "-I",
+        lambda flag: flag[:3] != "-Wp",
+        lambda flag: flag[:2] != "-D",
+        lambda flag: flag != "-include",
+        lambda flag: r"./" not in flag,  # no relative pathing allowed!
+        lambda flag: not flag.startswith("-Werror"),
+    ]
+
+    cleaned = cc_invocation[::]
+    for rule in rules:
+        cleaned = filter(rule, cleaned)
+
+    cleaned = list(cleaned)
+
+    ic(cleaned)
+
+    return cleaned
+
+
+def main() -> None:
+    cli_args: argparse.Namespace = parse_cli_args()
+
+    ic(cli_args)
+    make_stdout = initial_make(
+        build_command=cli_args.build_command,
+        target=cli_args.target,
+        path_to_linux=cli_args.path_to_linux,
+        force_rm=cli_args.force_rm_existing_target,
+    )
+
+    ic(make_stdout)
+
+    cc_invocation = get_compiler_invocation(make_stdout)
+    path_to_dot_i_file = make_dot_i_file(
+        cc_invocation, path_to_linux=cli_args.path_to_linux, output_dir=cli_args.output
+    )
+
+    clean_cc_invocation = get_clean_compiler_invocation(cc_invocation)
+    clean_cc_invocation[-1] = str(
```

Comment:
As mentioned before, `subprocess.run()` can handle path-like objects, so the cast to `str` should not be necessary.